### PR TITLE
[DYNAREC_RV64] Added support for xtheadba

### PR DESCRIPTION
**Pull request by ksco**

### CODE REVIEW

**Comment by ksco** [comment]
Dont forget to squash ;-)

**Comment by ksco** [comment]
```
                if (rv64_zba) {
                    SHxADD(x3, x1, 2+rex.w, wback);
                } else if (rv64_xtheadba) {
                    TH_ADDSL(x3, wback, x1, 2+rex.w);
                } else {
                    SLLI(x1, x1, 2+rex.w);
                    ADD(x3, wback, x1);
                }
```

This pattern is common, maybe a macro?

**Comment by ptitSeb** [comment]
> ```
>                 if (rv64_zba) {
>                     SHxADD(x3, x1, 2+rex.w, wback);
>                 } else if (rv64_xtheadba) {
>                     TH_ADDSL(x3, wback, x1, 2+rex.w);
>                 } else {
>                     SLLI(x1, x1, 2+rex.w);
>                     ADD(x3, wback, x1);
>                 }
> ```
> 
> This pattern is common, maybe a macro?

Yeah, if you can do a macro for this one; like `SLxADD(r1, r2, imm3)`?

**Comment by ksco** [comment]
Done, but why CI is failing, it passed on my machine.

**Comment by ptitSeb** [comment]
> Done, but why CI is failing, it passed on my machine.

I don't know, but that seems like a leggit error.

**Comment by ksco** [comment]
> > Done, but why CI is failing, it passed on my machine.
> 
> I don't know, but that seems like a leggit error.

Yeah it's because qemu and my machine have different extensions, fixing... We should run the tests multi times with different extension sets...

File: src/dynarec/rv64/rv64_emitter.h [hunk 1]
```diff
@@ -310,6 +310,16 @@ f28–31  ft8–11  FP temporaries                  Caller
 // rd = rs1 + imm12
 #define ADDIz(rd, rs1, imm12)       EMIT(I_type((imm12)&0b111111111111, rs1, 0b000, rd, rex.is32bits?0b0011011:0b0010011))
 
+// rd = rs1 + (rs2 << imm2), rs2 might be used as scratch.
+#define ADDSL(rd, rs1, rs2, imm2) if (rv64_zba) {   \
+        SHxADD(rd, rs2, imm2, rs1);                 \
+    } else if (rv64_xtheadba) {                     \
+        TH_ADDSL(rd, rs1, rs2, imm2);               \
+    } else {                                        \
+        SLLI(rs2, rs2, imm2);                       \
```
Comment (ptitSeb) [inline src/dynarec/rv64/rv64_emitter.h]: this does not reflect those line: `SLLI(ret, xRAX+sib_reg, (sib>>6));` for example. I guess that's the cause of the CI issues.

Comment (ksco) [reply src/dynarec/rv64/rv64_emitter.h]: Yeah yeah yeah, I already fixed it, you seem reviewing an old version.

Comment (ptitSeb) [reply src/dynarec/rv64/rv64_emitter.h]: multitasking... I'm lagging ;)

**Pull request merged**